internal/client/state-db/noncanonical.go

```diff
@@ -0,0 +1,670 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package statedb
+
+import (
+	"fmt"
+	"log"
+	"math/bits"
+
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/gammazero/deque"
+)
+
+var lastCanonical = []byte("last_canonical")
+
+const maxBlocksPerLevel uint64 = 32
+
+// NonCanonicalOverlay maintains trees of block overlays and allows discarding trees/roots.
+// The overlays are added in `Insert` and removed in `Canonicalize`.
+type NonCanonicalOverlay[BlockHash Hash, Key Hash] struct {
+	lastCanonicalized *hashBlock[BlockHash]
+	levels            deque.Deque[overlayLevel[BlockHash, Key]]
+	parents           map[BlockHash]BlockHash
+	values            map[Key]struct {
+		count uint32
+		value DBValue
+	}
+	// would be deleted but kept around because block is pinned, ref counted.
+	pinned           map[BlockHash]uint32
+	pinnedInsertions map[BlockHash]struct {
+		keys  []Key
+		count uint32
+	}
+	pinnedCanonicalized []BlockHash
+}
+
+type hashBlock[BlockHash Hash] struct {
+	Hash  BlockHash
+	Block uint64
+}
+
+// NewNonCanonicalOverlay is constructor for NonCanonicalOverlay
+func NewNonCanonicalOverlay[BlockHash Hash, Key Hash](db MetaDB) (NonCanonicalOverlay[BlockHash, Key], error) {
```

Comment:
Why aren't we returning a pointer?

```suggestion
func NewNonCanonicalOverlay[BlockHash Hash, Key Hash](db MetaDB) (*NonCanonicalOverlay[BlockHash, Key], error) {
```

Reply:
This will actually be stored in the `StateDB` struct which will be addressable.